File: 2023/python/day01/main.py
import re
import sys


def get_code(line: str):
    subs ={
        'one': '1',
        'two': '2',
        'three': '3',
        'four': '4',
        'five': '5',
        'six': '6',
        'seven': '7',
        'eight': '8',
        'nine': '9',
    }

    first_match, last_match = None, None

    expr = r'[1-9]|one|two|three|four|five|six|seven|eight|nine'
    first_match = re.search(expr, line)

    for i in range(0, len(line)):
        last_match = re.search(expr, line[len(line)-1-i:])
        if last_match is not None:
            break

    if first_match is None and last_match is None:
        return 0
    
    first_digit = subs[first_match.group()] if first_match.group() in subs else first_match.group()
    last_digit = subs[last_match.group()] if last_match.group() in subs else last_match.group()
    
    return int(first_digit + last_digit)


def get_calibration_code(line: str):
    digits = [d for d in line if d.isdigit()]

    if len(digits) == 0:
        return 0

    return int(digits[0] + digits[-1])


def solve_puzzle(data: str):
    result1 = sum(get_calibration_code(line) for line in data.splitlines())
    print("Part one answer:", result1)

    result2 = sum(get_code(line) for line in data.splitlines())
    print("Part two answer:", result2)


if __name__ == "__main__":
    with open(sys.argv[1], 'r') as f:
        puzzle_input = f.read()
    solve_puzzle(puzzle_input)

File: 2023/golang/day01/main.go
package main

import (
	"aoc-2023-go/internal/input"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	data := input.ReadPuzzle(os.Args[1])
	fmt.Printf("Part 1 answer: %d\n", sum(data, firstDigit))
	fmt.Printf("Part 2 answer: %d\n", sum(data, firstDigitOrWord))
}

func sum(input string, f func(string) int) int {
	total := 0

	for _, line := range strings.Split(input, "\n") {
		for i := 0; i < len(line); i++ {
			if d := f(line[i:]); d > 0 {
				total += d * 10
				break
			}
		}
		for i := len(line) - 1; i >= 0; i-- {
			if d := f(line[i:]); d > 0 {
				total += d
				break
			}
		}
	}
	return total
}

func firstDigit(s string) int {
	if s[0] >= '0' && s[0] <= '9' {
		val, _ := strconv.Atoi(string(s[0]))
		return val
	}
	return 0
}

func firstDigitOrWord(s string) int {
	if d := firstDigit(s); d > 0 {
		return d
	}
	words := [9]string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
	for i, w := range words {
		if strings.HasPrefix(s, w) {
			return i + 1
		}
	}
	return 0
}
